synapse/storage/databases/main/events.py

```diff
@@ -421,61 +414,69 @@ async def _persist_events_and_state_updates(
-                            continue
-
-                    if sender_level >= redact_level:
-                        await self.db_pool.simple_upsert(
-                            "room_ban_redactions",
-                            {"room_id": event.room_id, "user_id": event.state_key},
-                            {
-                                "redacting_event_id": event.event_id,
-                                "redact_end_ordering": None,
-                            },
-                            {
-                                "room_id": event.room_id,
-                                "user_id": event.state_key,
-                                "redacting_event_id": event.event_id,
-                                "redact_end_ordering": None,
-                            },
-                        )
 
-                        # normally the cache entry for a redacted event would be invalidated
-                        # by an arriving redaction event, but since we are not creating redaction
-                        # events we invalidate manually
-                        self.store._invalidate_local_get_event_cache_room_id(
-                            event.room_id
-                        )
+                    # normally the cache entry for a redacted event would be invalidated
+                    # by an arriving redaction event, but since we are not creating redaction
+                    # events we invalidate manually
+                    self.store._invalidate_local_get_event_cache_room_id(event.room_id)
+
+                    self.store._invalidate_async_get_event_cache_room_id(event.room_id)
 
             if new_forward_extremities:
                 self.store.get_latest_event_ids_in_room.prefill(
                     (room_id,), frozenset(new_forward_extremities)
                 )
 
+    async def _can_sender_redact(self, event: EventBase) -> bool:
+        state_filter = StateFilter.from_types([(EventTypes.PowerLevels, "")])
+        state = await self.store.get_partial_filtered_current_state_ids(
+            event.room_id, state_filter
+        )
+        pl_id = state[(EventTypes.PowerLevels, "")]
+        pl_event = await self.store.get_event(pl_id)
+        if pl_event:
+            sender_level = pl_event.content.get("users", {}).get(event.sender)
+            if sender_level is None:
+                sender_level = pl_event.content.get("users_default", 0)
+
+            redact_level = pl_event.content.get("redact")
+            if not redact_level:
+                redact_level = pl_event.content.get("events_default", 0)
+
+            room_redaction_level = pl_event.content.get("events", {}).get(
+                "m.room.redaction"
+            )
+            if room_redaction_level:
+                if sender_level < room_redaction_level:
+                    return False
+
+            if sender_level >= redact_level:
+                return True
+
+        return False
```

Comment:
nit: reduce indentation

```suggestion
        if pl_event is None:
            return False

        sender_level = pl_event.content.get("users", {}).get(event.sender)
        if sender_level is None:
            sender_level = pl_event.content.get("users_default", 0)

        redact_level = pl_event.content.get("redact")
        if not redact_level:
            redact_level = pl_event.content.get("events_default", 0)

        room_redaction_level = pl_event.content.get("events", {}).get(
            "m.room.redaction"
        )
        if room_redaction_level:
            if sender_level < room_redaction_level:
                return False

        if sender_level >= redact_level:
            return True

        return False
```

Reply:
In fact, if `pl_event` doesn't exist, we should fall back to the [default values](https://spec.matrix.org/v1.15/client-server-api/#mroompower_levels) for `m.room.power_levels`.

Reply:
If `pl_event` doesn't exist, and we fall back to the default values, redact would be 50 and the sender level would be 0, so it would always be false/fail?